Library/Homebrew/rubocops/cask/no_overrides.rb

```diff
@@ -0,0 +1,53 @@
+# typed: true
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Cask
+      class NoOverrides < Base
+        extend T::Sig
+        include CaskHelp
+
+        ON_SYSTEM_METHODS = RuboCop::Cask::Constants::ON_SYSTEM_METHODS
+        MESSAGE = <<~EOS
+          Do not use a top-level `%<stanza>s` stanza as the default. Add it to an `on_{system}` block instead.
+          Use `:or_older` or `:or_newer` to specify a range of macOS versions.
+        EOS
+
+        def on_cask(cask_block)
+          return if (cask_stanzas = cask_block.toplevel_stanzas).empty?
+          # Skip if there are no `on_*` blocks.
+          return unless (on_blocks = cask_stanzas.select { |s| ON_SYSTEM_METHODS.include?(s.stanza_name) }).any?
+
+          cask_stanzas.each do |stanza|
+            # Skip if the stanza is itself an `on_*` block.
+            next if ON_SYSTEM_METHODS.include?(stanza.stanza_name)
+            # Skip if the stanza we detect is already in an `on_*` block.
+            next if stanza.parent_node.block_type? && ON_SYSTEM_METHODS.include?(stanza.parent_node.method_name)
```

Comment:
Do we need this check? Doesn't `cask_block.toplevel_stanzas` return only the top-level stanzas?

Reply:
We do need this because `toplevel_stanzas` includes `version` and it's fine if the `version` stanza in the Cask is in an `on_*` block already.

(To be sure, I removed this `next` and the "RuboCop::Cop::Cask::NoOverrides when there are no top-level standalone stanzas does not report any offenses" test failed.)

Reply:
> We do need this because `toplevel_stanzas` includes `version` and it's fine if the `version` stanza in the Cask is in an `on_*` block already.

So `cask_block.toplevel_stanzas` includes `version` if the cask only contains the following?

```ruby
on_mojave :or_later do
  version :latest
end
```

If that is the case, that doesn't seem right.

Reply:
This is because `toplevel_stanzas` is https://github.com/Homebrew/brew/blob/ffa7ee01126c3ae59c80e684550702ccaa5aeba9/Library/Homebrew/rubocops/cask/ast/cask_block.rb#L38 and that `toplevel_stanza?` is defined as https://github.com/Homebrew/brew/blob/ffa7ee01126c3ae59c80e684550702ccaa5aeba9/Library/Homebrew/rubocops/cask/ast/stanza.rb#L55-L57. So it _does_ support one level of nesting. That explains why I had to have this condition to satisfy that test.